base/sparse/sparsematrix.jl

```diff
@@ -613,115 +613,326 @@ function sparse(B::Bidiagonal)
+    _distributevals_halfperm!(X, C, p, identity)
+    return X
+end
+"""
+    unchecked_aliasing_permute!{Tv,Ti<:Integer}(A::SparseMatrixCSC{Tv,Ti},
+        p::AbstractVector, q::AbstractVector, C::SparseMatrixCSC{Tv,Ti},
+        workcolptr::Vector{Ti})
+
+See [`permute!`](:func:`Base.SparseArrays.permute!`) for basic usage. Parent of `permute!`
+methods operating on `SparseMatrixCSC`s where the source and destination matrices are the
+same. See [`unchecked_noalias_permute!`](:func:`Base.SparseArrays.unchecked_noalias_permute`)
+for additional information; these methods are identical but for this method's requirement of
+the additional `workcolptr`, `length(workcolptr) >= A.n + 1`, which enables efficient
+handling of the source-destination aliasing.
+"""
+function unchecked_aliasing_permute!{Tv,Ti<:Integer}(A::SparseMatrixCSC{Tv,Ti},
+        p::AbstractVector, q::AbstractVector, C::SparseMatrixCSC{Tv,Ti},
+        workcolptr::Vector{Ti})
+    halfperm!(C, A, q)
+    _computecolptrs_permute!(A, A, q, workcolptr)
+    _distributevals_halfperm!(A, C, p, identity)
+    return A
+end
+"""
+Helper method for `unchecked_noalias_permute!` and `unchecked_aliasing_permute!`.
+Computes `PAQ`'s column pointers, storing them shifted one position forward in `X.colptr`;
+`_distributevals_halfperm!` fixes this shift. Saves some work relative to
+`_computecolptrs_halfperm!` as described in `uncheckednoalias_permute!`'s documentation.
+"""
+function _computecolptrs_permute!{Tv,Ti<:Integer}(X::SparseMatrixCSC{Tv,Ti},
+        A::SparseMatrixCSC{Tv,Ti}, q, workcolptr::Vector{Ti})
+    # Compute `A[p,q]`'s col counts. Store shifted forward one position in workcolptr.
+    @inbounds for k in 1:A.n
+        workcolptr[k+1] = A.colptr[q[k] + 1] - A.colptr[q[k]]
+    end
+    # Compute `A[p,q]`'s colptrs. Store shifted forward one position in X.colptr.
+    X.colptr[1] = 1
+    countsum = 1
+    @inbounds for k in 2:(X.n + 1)
+        overwritten = workcolptr[k]
+        X.colptr[k] = countsum
+        countsum += overwritten
     end
-
-    # Tracking write positions in Ccolptr as in the last block fixes the colptr shift,
-    # but the first colptr remains incorrect
-    Ccolptr[1] = 1
-
-    C
 end
-transpose!{Tv,Ti}(C::SparseMatrixCSC{Tv,Ti}, A::SparseMatrixCSC{Tv,Ti}) = qftranspose!(C, A, 1:A.n, identity)
-ctranspose!{Tv,Ti}(C::SparseMatrixCSC{Tv,Ti}, A::SparseMatrixCSC{Tv,Ti}) = qftranspose!(C, A, 1:A.n, conj)
-"See `qftranspose!`" ftranspose!{Tv,Ti}(C::SparseMatrixCSC{Tv,Ti}, A::SparseMatrixCSC{Tv,Ti}, f) = qftranspose!(C, A, 1:A.n, f)
 
 """
-    qftranspose{Tv,Ti}(A::SparseMatrixCSC{Tv,Ti}, q::AbstractVector, f)
+Helper method for `permute` and `permute!` methods operating on `SparseMatrixCSC`s.
+Checks compatability of source argument `A`, row-permutation argument `p`, and
+column-permutation argument `q`.
+"""
+function _checkargs_sourcecompatperms_permute!{Tv,Ti<:Integer}(A::SparseMatrixCSC{Tv,Ti},
+         p::AbstractVector, q::AbstractVector)
```

Comment:
should this also check that the inputs are permutations? if they have repeats or out-of-range values would bad things happen?